### DISCUSSION TITLE
Use of custom fonts with OpenType feature

### DISCUSSION THREAD

**Discussion by C-Mon83** [RudolphRiedel/FT800-FT813#144] [Q&A]
Hello Rudolph

I am currently working with a 4.3” EVE4 display from Riverdi [RVT43HLBFWN00](https://www.mouser.ch/datasheet/2/1021/DS_RVT43HLBFWN00_Rev_1_6-3137769.pdf). To check the compatibility of the display for other national languages, I have output a sample text with Indian characters. The text on the display has shown that certain characters are not displayed correctly. I used the font [MUKTI](https://fontlibrary.org/de/font/mukti). I have used the sample text from the specified website as a reference text.

![grafik](https://github.com/user-attachments/assets/70e757b9-8f02-405a-a568-d249a959ea36)

However, the following text is shown on the display and in the EVE Screen Editor as well:

![grafik](https://github.com/user-attachments/assets/f343c3b2-de41-42f4-b175-cc9f94ddf363)

What I noticed directly in the comparison are these dotted circles. After analyzing this font on [FontDrop!](https://fontdrop.info), I found out that this must be the OpenType feature.

![grafik](https://github.com/user-attachments/assets/57611411-2e45-4e5f-9cc3-27cb45715779)

I am now wondering whether the EVE display is able to handle this OpenType feature at all, or whether there are fonts for the Indian language that can manage without this OpenType feature.

**Implementation**

I downloaded the font from the URL above and prepared it in the EAB tool as follows.

![grafik](https://github.com/user-attachments/assets/9d5d76bc-7e7a-4737-bfa7-1e73eca65ccb)

The input character preview looked like this:

![grafik](https://github.com/user-attachments/assets/fc6b01de-04ca-49bf-a688-a403b44c8d94)

I created the Flash image as follows:

![grafik](https://github.com/user-attachments/assets/8714168a-8640-4a55-9050-9072c643a176)

I then compressed the image with the zopfli library.

![grafik](https://github.com/user-attachments/assets/68f300f9-cbe3-4007-8ef8-240440bb62c8)

I converted the compressed image file into a C file for integration into the software.

![grafik](https://github.com/user-attachments/assets/ef2588d3-626a-4f34-a352-22b23c86dc0a)

The source code to load and display the custom font looks like this:

`

            EVE_cmd_inflate(0, flashImage, sizeof(flashImage)); /* de-compress flash-image to RAM_G */
            uint32_t datasize = EVE_cmd_getptr(); /* we unpacked to RAM_G address 0x0000, so the first address after the unpacked data also is the size */

            EVE_cmd_flashupdate(0,0,4096); /* write blob first */
            if (E_OK == EVE_init_flash())
            {
                EVE_cmd_flashupdate(0,0,(datasize|4095)+1); /* size must be a multiple of 4096, so set the lower 12 bits and add 1 */
            }

            if (E_OK == EVE_init_flash())
            {
                EVE_cmd_flashread(0, 180928, 1408); /* copy .xfont from FLASH to RAM_G, offset and length are from the .map file */

                // *******************************************************************
                EVE_cmd_dl(CMD_DLSTART);
                EVE_cmd_dl(DL_CLEAR_COLOR_RGB | 0xFFFFFF);
                EVE_cmd_dl(DL_CLEAR | CLR_COL | CLR_STN | CLR_TAG);

                EVE_color_rgb(0x000000);
                EVE_cmd_setfont2(16, 0, 0);
                EVE_cmd_text(EVE_HSIZE/2, EVE_VSIZE/2, 16, EVE_OPT_CENTER, "মুক্তি মুক্ত বাংলা ফন্ট");
                EVE_cmd_text(EVE_HSIZE/2, EVE_VSIZE/2+30, 16, EVE_OPT_CENTER, "Hello World!");

                EVE_cmd_dl(DL_DISPLAY); /* mark the end of the display-list */
                EVE_cmd_dl(CMD_SWAP); /* make this list active */
                // *******************************************************************
`

I also tested the output in the EVE Screen Editor as follows:

![grafik](https://github.com/user-attachments/assets/92fb1396-90c8-4ebc-bbc9-7ceffc4d566f)

Basically I think that it is possible to show characters of all languages that exist in this world on this display. I also thought that the use of a Unicode character set would make this possible without any problems.

Have I overlooked something or do I have to use a different font to be able to output the selected sample text?

Kind regards
Simon

**Comment by RudolphRiedel** [comment]
There just was an issue that got resolved by using the latest EAB 2.12.2 (and this issue is not).
I converted mukti.otf with EAB 2.12.2 to have a look - not that I have any idea what these characters mean.
The number of converted glyphs is the same, only 187.

When I put the font into https://fontdrop.info/ I get that 601 glyphs are in the font and "12 OpenType features were detected in the font" - no idea what a "feature" is even supposed to be in the OTF context.

The preview image from EAB has this:
![grafik](https://github.com/user-attachments/assets/909b7114-2441-4a6f-8de8-cba64fcfe337)

So the circles get embedded into the glyphs, this is a conversion issue.

I also tried four different only OTF to TTF conversion services, three of the generated .ttf still show these OpenType "features" in FontDrop.
And all four .ttf get converted to only 187 glyphs in EAB with the circles embedded.

No idea what this is exactly, but the issue is not with the EVE chip.

**Reply by C-Mon83** [reply]
Hello Rudolph

Thank you for the quick reply and the tests with EAB version 2.12.2. I will also download the latest version.

As far as I understand, the Unicode table created with EAB is correct and the characters are identical to the original table ([Unicode Block "Bengali"](https://www.unicode.org/charts/PDF/U0980.pdf)). From my point of view, combining diacritical marks during parsing does not work ([Bengali Diacritcs](https://en.wikibooks.org/wiki/Bengali/Script/Diacritics)).

I have written to Bridgetek support about this. I will update my post as soon as I have an answer.

Kind regards
Simon
